src/goto-instrument/contracts/havoc_assigns_clause_targets.cpp

```diff
@@ -0,0 +1,252 @@
+/*******************************************************************\
+
+Module: Havoc multiple and possibly dependent targets simultaneously
+
+Author: Remi Delmas, delmasrd@amazon.com
+
+\*******************************************************************/
+
+/// \file
+/// Havoc multiple and possibly dependent targets simultaneously
+
+#include "havoc_assigns_clause_targets.h"
+
+#include <map>
+
+#include <langapi/language_util.h>
+
+#include <util/c_types.h>
+#include <util/format_expr.h>
+#include <util/format_type.h>
+#include <util/pointer_expr.h>
+#include <util/pointer_offset_size.h>
+#include <util/pointer_predicates.h>
+#include <util/std_code.h>
+
+#include "utils.h"
+
+static const char ASSIGNS_CLAUSE_REPLACEMENT_TRACKING[] =
+  " (assigned by the contract of ";
+
+static irep_idt make_tracking_comment(
+  const exprt &target,
+  const irep_idt &function_id,
+  const namespacet &ns)
+{
+  return from_expr(ns, target.id(), target) +
+         ASSIGNS_CLAUSE_REPLACEMENT_TRACKING + id2string(function_id) + ")";
+}
+
+bool is_assigns_clause_replacement_tracking_comment(const irep_idt &comment)
+{
+  return id2string(comment).find(ASSIGNS_CLAUSE_REPLACEMENT_TRACKING) !=
+         std::string::npos;
+}
+
+/// Returns a pointer expression to the start address of the target
+///
+/// - If the target is a `pointer_object(p)` expression,
+///   return a pointer to the same object with offset zero
+/// - else, if the target is a sized and assignable expression,
+///   return its address
+/// - else trigger an error.
+///
+static exprt base_pointer(
+  const exprt &target,
+  // context parameters
+  const namespacet &ns)
+{
+  if(target.id() == ID_pointer_object)
+  {
+    const auto &ptr = target.operands().front();
+    // bring the offset back to zero
+    auto res = minus_exprt{ptr, pointer_offset(ptr)};
+    res.type() = ptr.type();
+    return std::move(res);
```

Comment:
Scratch my earlier comment, the following will actually do the right thing: `return minus_exprt{ptr, pointer_offset(ptr)};` (the type of the `minus_exprt` will be `ptr.type()` when constructed in this way).

Reply:
thanks